proposals/csharp-12.0/collection-expressions.md

```diff
@@ -219,7 +213,36 @@ ImmutableArray<int> ia =
 ## Construction
 [construction]: #construction
 
-*Give specific ordering for determining how to construct the constructible collection types.*
+A collection expression has a *known length* if the compile-time type of each *spread element* in the collection expression is [*countable*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-8.0/ranges.md#adding-index-and-range-support-to-existing-library-types).
+
+*Does *countable* include explicit implementations of `ICollection<T>.Count` and `ICollection.Count`? We should check for those interfaces as well.*
+
+If the target type is a *struct* or *class type* that implements `System.Collections.IEnumerable`, and the target type does not have a *[create method](#create-methods)*, the construction of the collection instance as follows:
+
+* If the collection expression does not contain *spread elements*, and the collection type contains an applicable constructor with a single `int capacity` parameter, the compiler *may* invoke that constructor with a non-zero capacity value. Otherwise, the constructor that is applicable with no arguments is invoked.
```

Comment:
this doesn't seem good to me either.  even if we have spread elements, we can call the capacity constructor.  The primary differentiation is if something has sttaically known length, or runtime known length.

Reply:
If we need to interleave element evaluation and `Add` calls, I don't think we can determine the length of all spread elements before the collection instance is constructed.

Reply:
> If we need to interleave element evaluation and Add calls

the question is whether we need to do that or not.  I'd say this makes sense for a collection *without* spreads (since we know syntactically the number of elemnets).  But if there's a computation of element-count, it probably makes sense to determine that first, create the collection with the right capacity, then add to that :)

Reply:
We should probably be consistent with interleaving element evaluation and `Add` calls regardless of whether there are spread elements. We could decide to evaluate all elements before adding any, although that would be a change from the classic *collection initializer* behavior.

Reply:
Yeah, let's bring to LDM.  there's pros/cons each way.  :)